src/java/org/jivesoftware/openfire/plugin/rest/controller/GroupController.java

```diff
@@ -136,10 +167,65 @@ public Group updateGroup(String groupName, GroupEntity groupEntity) throws Servi
         if (groupEntity != null && !groupEntity.getName().isEmpty()) {
             if (groupName.equals(groupEntity.getName())) {
                 try {
+                    final Collection<JID> newMembers = new HashSet<>();
+                    final Collection<JID> newAdmins = new HashSet<>();
+
+                    // Input validation.
                     group = GroupManager.getInstance().getGroup(groupName);
+                    for (final String newMember : groupEntity.getMembers()) {
+                        try {
+                            newMembers.add(newMember.contains("@") ? new JID(newMember) : XMPPServer.getInstance().createJID(newMember, null));
+                        } catch (IllegalArgumentException e) {
+                            throw new ServiceException("Cannot parse a member value as a JID: " + newMember, groupEntity.getName(), ExceptionType.ILLEGAL_ARGUMENT_EXCEPTION, Response.Status.BAD_REQUEST, e);
+                        }
+                    }
+                    for (final String newAdmin : groupEntity.getAdmins()) {
+                        try {
+                            newAdmins.add(newAdmin.contains("@") ? new JID(newAdmin) : XMPPServer.getInstance().createJID(newAdmin, null));
+                        } catch (IllegalArgumentException e) {
+                            throw new ServiceException("Cannot parse a admin value as a JID: " + newAdmin, groupEntity.getName(), ExceptionType.ILLEGAL_ARGUMENT_EXCEPTION, Response.Status.BAD_REQUEST, e);
+                        }
+                    }
+
                     group.setDescription(groupEntity.getDescription());
                     final String showInRoster = groupEntity.getShared() ? "onlyGroup" : "nobody";
                     group.getProperties().put("sharedRoster.showInRoster", showInRoster);
+
+                    // Correct the member-list that already is in the group to match the desired state.
+                    final Collection<JID> members = group.getMembers();
+                    final Iterator<JID> membersIterator = members.iterator();
+                    while (membersIterator.hasNext()) {
+                        final JID oldMember = membersIterator.next();
+                        if (newMembers.contains(oldMember)) {
+                            // Already exists. No need to add it again.
+                            newMembers.remove(oldMember);
+                        } else {
+                            // No longer exists. Remove from group.
+                            membersIterator.remove();
+                        }
+                    }
+                    for (final JID newMember : newMembers) {
+                        // Unsure if #addAll works for the specific Collection subclass that is used in the Group implementation. Let's add them one-by-one to be safe.
+                        members.add(newMember);
```

Comment:
Does it matter if the user is also an administrator? The UI doesn't let you add the user to "both groups" - it's represented as a checkbox.

Reply:
The specific Collection implementation that backs these collections (`org.jivesoftware.openfire.group.Group.MemberCollection`) will take care of this automatically.